### PR TITLE
Add dependabot scanner

### PR DESCRIPTION
**Pull request by zkdev**
**What this PR does / why we need it**:
Scans repositories from specified organization for `dependabot.yaml` to create a coverage report.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
None

**Release note**:

```improvement operator

```

### CODE REVIEW

**Review by ccwienk — CHANGES_REQUESTED** [review]

File: dso/dependabot.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    return github_org.repositories
+
+
+def _scan_repos_for_dependabot(
+    repos: typing.List[str],
+    host: str,
+) -> dso.model.DependabotCoverageReport:
+
+    clogger = dso.util.component_logger(__name__)
+    i = 1
```
Comment (ccwienk) [inline dso/dependabot.py:78]: nit: i is unused (de facto)

File: dso/dependabot.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        )
+        repo_report = dso.model.DependabotCoverageReportRepo(
+            dependabot=dependabot_found,
+            repo=repo
+        )
+        details.append(repo_report)
+        i += 1
+
+    return dso.model.DependabotCoverageReport(
+        coverage=0,
```
Comment (ccwienk) [inline dso/dependabot.py:103]: coverage=0? that cannot be what you intended. also: what is the semantics of this parameter?

File: dso/dependabot.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        details=details,
+        github=host,
+    )
+
+
+def _get_github_api(host):
```
Comment (ccwienk) [inline dso/dependabot.py:109]: nit: I would prefer to not name functions that are not simple getters (as in: return an attr (from an object), maybe doing some trivial operations on the attr) with a `get` prefix.

regardless if my nit, I think this convenience function should be moved to `ccc.github`, where it is obviously missing (in fact, I also needed such a function several times, but always was too lazy to implement one..)

+ please add a typehint for the return-value (and preferrably also the parameter - even though the latter is debatable, as it is just `str`..)

File: dso/dependabot.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    github_cfg = ccc.github.github_cfg_for_hostname(
+        cfg_factory=ci.util.ctx().cfg_factory(),
+        host_name=host,
+    )
+    return ccc.github.github_api(github_cfg)
+
+
+def _get_scoped_repos_for_org(
```
Comment (ccwienk) [inline dso/dependabot.py:117]: I do not think `scoped` is an appropriate term(*). In GitHub's terms, I "my" terms (as used elsewhere in cc-utils), I would tend to call `{org}/{repo}` a repo(sitory)-path. I think, github's API calls this the `full_name` of a repository. So repository-name might also be  adequate. or maybe qualified name.

(*) one could argue that the repo-name (w/o org) is only valid in the scope of its organisation. However, at least from my experience, I never saw the naming convention of calling a qualified name a scoped name (as opposed to canonical, qualified, or full name), and I find it unintuitive, at least (maybe it is not strictly "wrong").

File: dso/dependabot.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+) -> typing.List[str]:
+    clogger = dso.util.component_logger(__name__)
+    repos: typing.List[str] = []
+
+    clogger.info(f'org specified, scanning {org}')
+    host, org = org.split('/')
+    github_api = _get_github_api(host)
+    github_org = github_api.organization(org)
+    for repo in github_org.repositories():
+        repos.append(repo.full_name)
```
Comment (ccwienk) [inline dso/dependabot.py:128]: nit: this might be done more elegant using a list-comprehension

File: dso/dependabot.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+    return repos[:3]
```
Comment (ccwienk) [inline dso/dependabot.py:130]: why do you omit all but the first three repositories?

File: dso/dependabot.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+
+
+def _generate_reporting_tables(
+    report: dso.model.DependabotCoverageReport,
+    tablefmt: str,
+):
+    # monkeypatch: disable html escaping
+    tabulate.htmlescape = lambda x: x
```
Comment (ccwienk) [inline dso/dependabot.py:138]: nit: only required for html (+ TODO: check whether this has not already been fixed upstream :))

File: dso/dependabot.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+    tables.append(tabulate.tabulate(
+        headers=('Component', 'Dependabot'),
+        tabular_data=table_data,
+        tablefmt=tablefmt,
+        colalign=('left', 'center'),
+    ))
+
+    table_data = ((report.github, f'{report.coverage}%'),)
+    tables.append(tabulate.tabulate(
+        headers=('Overall Coverage', 'Percentage'),
```
Comment (ccwienk) [inline dso/dependabot.py:156]: nit: maybe express more explicit what "coverage" means (I assume it is the percentage of repositories that do use dependabot?)

File: dso/dependabot.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+    print('\n')
+
+
+def dependabot_coverage(
+    org: str,
+    outfile_path: str,
+):
+    repos = _get_scoped_repos_for_org(org=org)
+    host, _ = org.split('/')
+    report = _scan_repos_for_dependabot(repos=repos, host=host)
```
Comment (ccwienk) [inline dso/dependabot.py:185]: so you first create repository-objects (`_get_scoped_repos_for_org'), to then discard those objects again and only use their full-names, which you then use to create repository.-objects _again_? this sounds a bit cumbersome to me (also, it uses two requests for each repository, where we in fact only would need one :-/

File: dso/model.py [hunk 10]
```diff
@@ -17,3 +17,27 @@ class ScanArtifact:
...
+
+
+@dataclasses.dataclass(frozen=False)
+class DependabotCoverageReport:
+    coverage: float
+    github: str
+    details: typing.List[DependabotCoverageReportRepo]
+
+    def calculate_overall_percentage(self):
+        t = 0
```
Comment (ccwienk) [inline dso/model.py:35]: stylistic nit: this is a very cheap operation. rather either already do it in the c'tor, or (pbly better) do the calculation upfront, and pass-in the calculated values (-> only use this dataclass as a "struct"). I think it is a bad pattern to have a class-definition leading to objects that are useless w/o calling a special method once

**Review by ccwienk — CHANGES_REQUESTED** [review]
some more nits..

also, as a general remark, I suggest you either create a package `dependabot` on root-level of cc-utils, or a sub-package `sdo/dependabot`.

as for the repository-enumeration, I suggest so simply use the github3.py-api with less wrapping (both the org- and repository-objects it yields already have all the data you need).

roughly sketched, I would think we need one function that accepts a github3.py-api-repository-object, and returns a value indicating the dependabot-status (preferrably an enum), like so:

```
class DependebotStatus(enum.Enum):
 ENABLED = 'enabled'
 NOT_ENABLED = 'not_enabled'
 UNKNOWN = 'unknown' # not sure whether we need it. e.g. use it in case there was an error (say there is a dependabot-file, but it contained invalid yaml)

def dependabot_status(repository: github3.repos.repo.Repository):
 if not (files := retrieve_known_dependabot_files(repository)):
  return DependebotStatus.NOT_ENABLED
 # sanity-checks
 if len(files) > 1:
  # not sure - does dependabot allow this? - do something reasonable in this case..
  pass
 dependabot_file = files[0]
 if validate_dependabot_file(dependabot_file):
  return DependabotStatus.ENABLED
 else:
  # might also log a warning, ....
  return DependabotStatus.UNKNOWN

# using the function above, you might simply fill a mapping, like so:
repo_status = {
 repo: dependabot_status(repo)
 for repo in repositories
}
# where repositories is e.g. returned from a githu3.py-api-org-object

from such a dict, you can easily construct the reporting (read repo-full-name from repo (dict-key); dependabot-status is already the value..)

File: dependabot/status.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+
+def _validate_dependabot_file(
+    dependabot_file,
+) -> bool:
+
+    # TODO to be extended, e.g. scan for required attributes, see:
+    # https://docs.github.com/en/github/administering-a-repository/
+    # configuration-options-for-dependency-updates
+    content = dependabot_file.decoded.decode("utf-8")
```
Comment (ccwienk) [inline dependabot/status.py:36]: nit: by convention, we tend to prefer to use single-quotes in cc-utils

File: dependabot/status.py [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+        full_org_name=full_org_name,
+    )
+
+    with open(outfile_path, 'a') as f:
+        now = datetime.datetime.now()
+        f.write(f'{now.strftime("%d/%m/%Y %H:%M:%S")}\n')
+        for t in report_tables:
+            f.write(f'{t}\n\n')
+            print(f'\n{t}')
+        f.write('\n===================================================================\n\n')
```
Comment (ccwienk) [inline dependabot/status.py:113]: hint: you do know you can multiply strings? :-)

Comment (ccwienk) [reply dependabot/status.py:113]: also, I would like to advertise `textwrap.fill` (+ other functions^^)

File: dependabot/status.py [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+    full_org_name: str,
+):
+
+    report_tables = _generate_report_tables_from_repo_status(
+        repo_status=repo_status,
+        full_org_name=full_org_name,
+    )
+
+    with open(outfile_path, 'a') as f:
+        now = datetime.datetime.now()
```
Comment (ccwienk) [inline dependabot/status.py:108]: nit: how about isoformat()?

File: dependabot/status.py [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+        f.write(f'{now.strftime("%d/%m/%Y %H:%M:%S")}\n')
+        for t in report_tables:
+            f.write(f'{t}\n\n')
+            print(f'\n{t}')
+        f.write('\n===================================================================\n\n')
+    print('\n')
+
+
+def _calculate_coverage_rate(
```
Comment (ccwienk) [inline dependabot/status.py:117]: nit: rate should be between 0..1, should it not? what this function returns is probably rather called percentage, I think

File: dependabot/util.py [hunk 15]
```diff
@@ -0,0 +1,16 @@
+import github3.repos
+import typing
+
+import ccc.github
+
+
+def repositories_for_org(
+    full_org_name: str,  # e.g. github.com/gardener
```
Comment (ccwienk) [inline dependabot/util.py:8]: not sure I like `full_org_name` :-/
how about qualified_org, or github_org_url, or hostname_and_org?

File: dependabot/util.py [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+) -> typing.List[github3.repos.repo.ShortRepository]:
+
+    host, org = full_org_name.split('/')
```
Comment (ccwienk) [inline dependabot/util.py:11]: rather conditionally prepend a schema (if absent) + use urllib.parse.urlparse to parse into host and path?

**Review by ccwienk — CHANGES_REQUESTED** [review]

File: dependabot/status.py [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+    dependabot_file,
+) -> bool:
+
+    # TODO to be extended, e.g. scan for required attributes, see:
+    # https://docs.github.com/en/github/administering-a-repository/
+    # configuration-options-for-dependency-updates
+    content = dependabot_file.decoded.decode("utf-8")
+    try:
+        ci.util.load_yaml(content)
+    except (AttributeError,):
```
[the remark this answers is not in the document]
Comment (zkdev) [reply dependabot/status.py:39]: Was meant to be extended to catch more parsing errors, will adjust

File: dependabot/status.py [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+):
+
+    # According to https://docs.github.com/en/github/administering-a-repository/configuration-options
+    # -for-dependency-updates , there is only one valid path for the dependabot.yml.
+    try:
+        file = repository.file_contents(path='.github/dependabot.yml')
+        ci.util.info(f'dependabot in {repository.full_name}')
+        return file
+    except github3.exceptions.NotFoundError:
+        ci.util.info(f'dependabot not in {repository.full_name}')
```
[the remark this answers is not in the document]
Comment (zkdev) [reply dependabot/status.py:25]: True, will adjust

File: dependabot/status.py [hunk 19]
```diff
@@ -0,0 +1,142 @@
+import ci.util
```
[the remark this answers is not in the document]
Comment (zkdev) [reply dependabot/status.py:1]: Yea that was not meant to be ordered like this, will rearrange.

File: dependabot/model.py [hunk 20]
```diff
@@ -0,0 +1,15 @@
...
+
+class DependebotStatus(enum.Enum):
+    ENABLED = 'enabled'
+    NOT_ENABLED = 'not_enabled'
+    UNKNOWN = 'unknown'  # error case, e.g. dependabot.yaml found but faulty yaml
+
+
+@dataclasses.dataclass(frozen=True)
+class DependabotStatusForRepo:
+    repo: github3.repos.ShortRepository
```
[the remark this answers is not in the document]
Comment (zkdev) [reply dependabot/model.py:14]: sure, see: https://github.com/sigmavirus24/github3.py/blob/6572d7556e925f45b713bea3fb93f791e2ac9074/src/github3/models.py#L123

File: dependabot/status.py (hunk 11 above)
Comment (zkdev) [reply dependabot/status.py:36]: Absolutely.
Maybe we should extend our linter to check for unnecessary double quotation 🤔?

see: https://github.com/zheller/flake8-quotes

File: dependabot/util.py (hunk 15 above)
Comment (zkdev) [reply dependabot/util.py:8]: `github_org_url` sounds good

File: dependabot/model.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (zkdev) [reply dependabot/model.py:14]: the repo uri is hashed, see: https://github.com/sigmavirus24/github3.py/blob/6572d7556e925f45b713bea3fb93f791e2ac9074/src/github3/models.py#L43

File: dependabot/util.py (hunk 15 above)
Comment (zkdev) [inline dependabot/util.py:8]: here

**Comment by zkdev** [comment]
Apparently dependabot is enabled for the whole gardener org.
Will put this PR on hold.

**Comment by ccwienk** [comment]
closing, as discussed, as there is no need for dependabot-reportings (it is enabled for entire gh-orgs anyway)

@zkdev : sorry - my bad :-/

**Pull request closed without merging**